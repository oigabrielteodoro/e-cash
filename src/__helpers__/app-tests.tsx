import React, { useEffect, ReactElement, ReactNode } from 'react'
import { render as rtlRender } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {
  DefaultOptions,
  QueryCache,
  QueryClient,
  QueryClientProvider,
} from 'react-query'
import { MemoryRouter, Routes, Route } from 'react-router-dom'
import { generateId } from 'lib'
import nock from 'nock'

import { pipe } from 'fp-ts/function'
import { TaskEither, map, mapLeft } from 'fp-ts/TaskEither'

import { AppLayout, NotificationContainer } from 'ui'
import { clearToken, queryConfigDefault, setToken } from 'client'

type Options = {
  initialRoute?: string
  initialRoutePath?: string
  authenticated?: boolean
  queryClient?: QueryClient
  routePaths?: string[]
}

type Props = {
  children?: ReactNode
  client?: QueryClient
}

type Callback = (result: unknown) => unknown
type MapAll = (
  fn: Callback,
) => (data: TaskEither<unknown, unknown>) => TaskEither<unknown, unknown>

const queryCache = new QueryCache()
const defaultQueryClient = new QueryClient({
  defaultOptions: queryConfigDefault as DefaultOptions,
  queryCache: queryCache,
})

export function ReactQueryWrapper({
  children,
  client = defaultQueryClient,
}: Props) {
  return <QueryClientProvider client={client}>{children}</QueryClientProvider>
}

export function ReactRouterWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
}: Options & Pick<Props, 'children'>) {
  function MockedPage({ title }: { title: string }) {
    if (authenticated) {
      return (
        <AppLayout>
          <AppLayout.Content>Page {title}</AppLayout.Content>
        </AppLayout>
      )
    }

    return <h1>Page {title}</h1>
  }

  return (
    <MemoryRouter initialEntries={[initialRoute]}>
      <Routes>
        <Route path={initialRoute} element={children} />
        {routePaths?.map((routePath) => (
          <Route
            key={routePath}
            path={routePath}
            element={<MockedPage title={routePath} />}
          />
        ))}
      </Routes>
    </MemoryRouter>
  )
}

export function ReactHookWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
  client = defaultQueryClient,
}: Options & Props) {
  return (
    <ReactQueryWrapper client={client}>
      <ReactRouterWrapper
        initialRoute={initialRoute}
        routePaths={routePaths}
        authenticated={authenticated}
      >
        {children}
      </ReactRouterWrapper>
    </ReactQueryWrapper>
  )
}

function render(
  ui: ReactElement,
  {
    initialRoute = '/',
    queryClient = defaultQueryClient,
    authenticated = true,
    routePaths,
  }: Options = {},
) {
  function Wrapper({ children }: Props) {
    useEffect(() => {
      if (authenticated) {
        setToken(generateId())

        return () => {
          clearToken()
        }
      }
    }, [])

    return (
      <ReactQueryWrapper client={queryClient}>
        <ReactRouterWrapper
          initialRoute={initialRoute}
          routePaths={routePaths}
          authenticated={authenticated}
        >
          {children}
        </ReactRouterWrapper>

        <NotificationContainer />
      </ReactQueryWrapper>
    )
  }

  return {
    ...rtlRender(ui, {
      wrapper: Wrapper,
    }),
  }
}

export function unsafe<T>(value: unknown) {
  return value as T
}

export const mapAll: MapAll = (fn) => (data) => {
  return pipe(data, map(fn), mapLeft(fn))
}

export function getErrorMessage(errors: unknown): string {
  return Array.isArray(errors) ? errors[0].message : ''
}

export * from '@testing-library/react'
export {
  render,
  userEvent,
  nock,
  defaultQueryClient as queryClient,
  queryCache,
}
