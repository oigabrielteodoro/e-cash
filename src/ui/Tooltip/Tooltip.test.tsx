import React from 'react'
import { screen, render, userEvent, waitFor } from '__helpers__/app-tests'

import { Tooltip } from '.'

function MockedComponent() {
  return (
    <>
      <Tooltip message='Tooltip' position='top'>
        <button>Hover me Top</button>
      </Tooltip>
      <Tooltip message='Tooltip' position='bottom'>
        <button>Hover me Bottom</button>
      </Tooltip>
      <Tooltip message='Tooltip' position='right'>
        <button>Hover me Right</button>
      </Tooltip>
      <Tooltip message='Tooltip' position='left'>
        <button>Hover me Left</button>
      </Tooltip>
    </>
  )
}

describe('Tooltip', () => {
  it('should be able match snapshot', async () => {
    const { container } = render(<MockedComponent />)

    expect(
      screen.getByRole('button', {
        name: /hover me top/i,
      }),
    ).toBeInTheDocument()

    const button = screen.getByRole('button', {
      name: /hover me top/i,
    })

    userEvent.hover(button)

    await waitFor(() => expect(container).toMatchSnapshot())
  })

  it('should render on top position', async () => {
    render(
      <Tooltip message='Tooltip' position='top'>
        <button>Hover me</button>
      </Tooltip>,
    )

    expect(
      screen.getByRole('button', {
        name: /hover me/i,
      }),
    ).toBeInTheDocument()

    const button = screen.getByRole('button', {
      name: /hover me/i,
    })

    userEvent.hover(button)

    expect(await screen.findByText(/tooltip/i)).toBeInTheDocument()
  })

  it('should render on right position', async () => {
    render(
      <Tooltip message='Tooltip'>
        <button>Hover me</button>
      </Tooltip>,
    )

    expect(
      screen.getByRole('button', {
        name: /hover me/i,
      }),
    ).toBeInTheDocument()

    const button = screen.getByRole('button', {
      name: /hover me/i,
    })

    userEvent.hover(button)

    expect(await screen.findByText(/tooltip/i)).toBeInTheDocument()
  })

  it('should render on left position', async () => {
    render(
      <Tooltip message='Tooltip'>
        <button>Hover me</button>
      </Tooltip>,
    )

    expect(
      screen.getByRole('button', {
        name: /hover me/i,
      }),
    ).toBeInTheDocument()

    const button = screen.getByRole('button', {
      name: /hover me/i,
    })

    userEvent.hover(button)

    expect(await screen.findByText(/tooltip/i)).toBeInTheDocument()
  })

  it('should render on bottom position', async () => {
    render(
      <Tooltip message='Tooltip'>
        <button>Hover me</button>
      </Tooltip>,
    )

    expect(
      screen.getByRole('button', {
        name: /hover me/i,
      }),
    ).toBeInTheDocument()

    const button = screen.getByRole('button', {
      name: /hover me/i,
    })

    userEvent.hover(button)

    expect(await screen.findByText(/tooltip/i)).toBeInTheDocument()
  })
})
