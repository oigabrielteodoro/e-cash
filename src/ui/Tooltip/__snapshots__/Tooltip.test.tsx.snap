// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should be able match snapshot 1`] = `
.c2 {
  opacity: 0;
  visibility: hidden;
}

.c3 {
  position: absolute;
  color: #18171d;
  font-size: 1.2rem;
  font-weight: 500;
  background: #ffffff;
  padding: 1rem 1.6rem;
  border-radius: 0.5rem;
  white-space: nowrap;
  -webkit-transition: 300ms;
  transition: 300ms;
  border: 0.1rem solid #cfd7e3;
  text-align: center;
  z-index: 1;
  top: 0;
  left: 50%;
  -webkit-transform: translate(-50%,calc(-100% - 1.4rem));
  -ms-transform: translate(-50%,calc(-100% - 1.4rem));
  transform: translate(-50%,calc(-100% - 1.4rem));
}

.c3::before {
  content: '';
  position: absolute;
  border-style: solid;
  border-width: 0.6rem;
  border-color: transparent;
}

.c5 {
  position: absolute;
  color: #18171d;
  font-size: 1.2rem;
  font-weight: 500;
  background: #ffffff;
  padding: 1rem 1.6rem;
  border-radius: 0.5rem;
  white-space: nowrap;
  -webkit-transition: 300ms;
  transition: 300ms;
  border: 0.1rem solid #cfd7e3;
  text-align: center;
  z-index: 1;
  left: 50%;
  bottom: 0;
  -webkit-transform: translate(-50%,calc(100% + 1.4rem));
  -ms-transform: translate(-50%,calc(100% + 1.4rem));
  transform: translate(-50%,calc(100% + 1.4rem));
}

.c5::before {
  content: '';
  position: absolute;
  border-style: solid;
  border-width: 0.6rem;
  border-color: transparent;
}

.c7 {
  position: absolute;
  color: #18171d;
  font-size: 1.2rem;
  font-weight: 500;
  background: #ffffff;
  padding: 1rem 1.6rem;
  border-radius: 0.5rem;
  white-space: nowrap;
  -webkit-transition: 300ms;
  transition: 300ms;
  border: 0.1rem solid #cfd7e3;
  text-align: center;
  z-index: 1;
  top: 50%;
  right: 0;
  -webkit-transform: translate(calc(100% + 1.6rem),-50%);
  -ms-transform: translate(calc(100% + 1.6rem),-50%);
  transform: translate(calc(100% + 1.6rem),-50%);
}

.c7::before {
  content: '';
  position: absolute;
  border-style: solid;
  border-width: 0.6rem;
  border-color: transparent;
}

.c9 {
  position: absolute;
  color: #18171d;
  font-size: 1.2rem;
  font-weight: 500;
  background: #ffffff;
  padding: 1rem 1.6rem;
  border-radius: 0.5rem;
  white-space: nowrap;
  -webkit-transition: 300ms;
  transition: 300ms;
  border: 0.1rem solid #cfd7e3;
  text-align: center;
  z-index: 1;
  top: 50%;
  left: 0;
  -webkit-transform: translate(calc(-100% - 1.6rem),-50%);
  -ms-transform: translate(calc(-100% - 1.6rem),-50%);
  transform: translate(calc(-100% - 1.6rem),-50%);
}

.c9::before {
  content: '';
  position: absolute;
  border-style: solid;
  border-width: 0.6rem;
  border-color: transparent;
}

.c4 {
  position: absolute;
  background: #ffffff;
  height: 1rem;
  width: 1rem;
  border: 0.1rem solid #cfd7e3;
  z-index: 1;
  bottom: -0.8rem;
  top: -1.9rem;
  left: 50%;
  -webkit-transform: translateX(-50%) rotate(-45deg);
  -ms-transform: translateX(-50%) rotate(-45deg);
  transform: translateX(-50%) rotate(-45deg);
  border-top-color: transparent;
  border-right-color: transparent;
}

.c6 {
  position: absolute;
  background: #ffffff;
  height: 1rem;
  width: 1rem;
  border: 0.1rem solid #cfd7e3;
  z-index: 1;
  bottom: -1.9rem;
  left: 50%;
  -webkit-transform: translateX(-50%) rotate(-45deg);
  -ms-transform: translateX(-50%) rotate(-45deg);
  transform: translateX(-50%) rotate(-45deg);
  border-bottom-color: transparent;
  border-left-color: transparent;
}

.c8 {
  position: absolute;
  background: #ffffff;
  height: 1rem;
  width: 1rem;
  border: 0.1rem solid #cfd7e3;
  z-index: 1;
  top: 50%;
  left: calc(100% + 1.2rem);
  -webkit-transform: translateY(-50%) rotate(-45deg);
  -ms-transform: translateY(-50%) rotate(-45deg);
  transform: translateY(-50%) rotate(-45deg);
  border-bottom-color: transparent;
  border-right-color: transparent;
}

.c10 {
  position: absolute;
  background: #ffffff;
  height: 1rem;
  width: 1rem;
  border: 0.1rem solid #cfd7e3;
  z-index: 1;
  top: 50%;
  right: calc(100% + 1.2rem);
  -webkit-transform: translateY(-50%) rotate(-45deg);
  -ms-transform: translateY(-50%) rotate(-45deg);
  transform: translateY(-50%) rotate(-45deg);
  border-top-color: transparent;
  border-left-color: transparent;
}

.c0 {
  position: relative;
}

.c0:hover .c1 {
  opacity: 1;
  visibility: visible;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1 c2"
    >
      <span
        class="c3"
      >
        Tooltip
      </span>
      <div
        class="c4"
      />
    </div>
    <button>
      Hover me Top
    </button>
  </div>
  <div
    class="c0"
  >
    <div
      class="c1 c2"
    >
      <span
        class="c5"
      >
        Tooltip
      </span>
      <div
        class="c6"
      />
    </div>
    <button>
      Hover me Bottom
    </button>
  </div>
  <div
    class="c0"
  >
    <div
      class="c1 c2"
    >
      <span
        class="c7"
      >
        Tooltip
      </span>
      <div
        class="c8"
      />
    </div>
    <button>
      Hover me Right
    </button>
  </div>
  <div
    class="c0"
  >
    <div
      class="c1 c2"
    >
      <span
        class="c9"
      >
        Tooltip
      </span>
      <div
        class="c10"
      />
    </div>
    <button>
      Hover me Left
    </button>
  </div>
</div>
`;
